src/ir/principal-type.cpp

```diff
@@ -0,0 +1,1061 @@
+      o << " ";
+    }
+    print(o, *it);
+  }
+  if (type.unreachable) {
+    o << "]~>[";
+  } else {
+    o << "]->[";
+  }
+  for (auto it = type.results.begin(); it != type.results.end(); ++it) {
+    if (it != type.results.begin()) {
+      o << " ";
+    }
+    print(o, *it);
+  }
+  o << "]";
+}
+
+} // anonymous namespace
+
+bool PrincipalType::compose(const PrincipalType& next) {
+  assert(&next != this);
+
+  // Renumber the variables in this type to ensure they're distinct from the
+  // variables in `next`. `counts` will end up with the total counts across both
+  // types.
+  VarCounts counts;
+  countVars(counts, next);
+  renumberVars(counts, *this);
+
+  // Match up the provided and required types, collecting values for and
+  // unifying variables.
+  Index numProvided = results.size();
+  Index numRequired = next.rparams.size();
+  Index numMatched = std::min(numProvided, numRequired);
+  VarAssignments assignments;
+  for (Index i = 0; i < numMatched; ++i) {
+    const auto& provided = results[numProvided - i - 1];
+    const auto& required = next.rparams[i];
+    if (!match(assignments, provided, required)) {
+      // Undo our previous renumbering of indices.
+      canonicalizeVars(*this);
+      return false;
+    }
+  }
+
+  // The matched provided and required types are consistent, so from this point
+  // we know the composition will succeed and we can start mutating things. The
+  // matched provided types are annihilated by the matched required types.
+  results.resize(numProvided - numMatched);
+
+  // Any extra required parameters are prepended to the current parameters (i.e.
+  // appended to the reversed list of parameters), unless they would be
+  // satisfied by popping from an unreachable stack instead.
+  if (numRequired > numMatched) {
+    if (unreachable) {
+      for (Index i = numMatched; i < numRequired; ++i) {
+        const auto& required = next.rparams[i];
+        [[maybe_unused]] bool assigned =
+          match(assignments, Type(Type::unreachable), required);
+        assert(assigned);
+      }
+    } else {
+      rparams.reserve(rparams.size() + (numRequired - numMatched));
+      for (Index i = numMatched; i < numRequired; ++i) {
+        rparams.push_back(next.rparams[i]);
+      }
+    }
+  }
+
+  if (next.unreachable) {
+    // The existing results do not make it past the following unreachable.
+    results.clear();
+    unreachable = true;
+  }
+
+  // Append the new results.
+  for (const VarType& result : next.results) {
+    results.push_back(result);
+  }
+
+  apply(assignments, *this);
+  // If a type variable was instantiated with bottom type (i.e. unreachable) due
+  // to popping from an unreachabile stack, the result type may end with some
+  // number of unreachables. This is nonsensical, since `unreachable` is not a
+  // concrete type. We could have alternatively left the variables
+  // uninstantiated, but it would have no corresponding introduction on the left
```

Comment:
```suggestion
  // uninstantiated, but it would have no corresponding introduction on the left
```
What does 'introduction' mean here?

Reply:
If we have `[]*>[] + [t0]->[t0]`, then naively we would end up with `[]~>[unreachable]`, but that's not valid. A simple fix would be to leave `t0` uninstantiated, giving us `[]*>[t0]`, but now we have an unconstrained variable on the right hand side that does not appear (i.e. is "not introduced") on the left hand side. The better result is `[]*>[]`, which is semantically identical.

Reply:
Oops, I forgot I was using `~>` rather than `*>` in the tests. Either way, it's a totally made up notation for a stack-polymorphic (i.e. unreachable) type.